azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+        )
+
+    def _send_register_request(self, event_data):
+        logger.info("Sending registration request from polling machine")
+        self._set_query_timer()
+
+        rid = str(uuid.uuid4())
+
+        publish_action = PublishAction(
+            constant.PUBLISH_TOPIC_REGISTRATION + rid, " ", callback=self.on_publish_completed
+        )
+        self._operations[rid] = publish_action
+        self._request_response_provider.send_request(publish_action)
+
+    def _query_operation_status(self, event_data):
+        logger.info("Querying operation status from polling machine")
+        self._set_query_timer()
+
+        rid = str(uuid.uuid4())
+        registration_result = event_data.args[0].args[0]
+
+        operation_id = registration_result.operation_id
+        publish_action = PublishAction(
+            constant.PUBLISH_TOPIC_QUERYING_I
+            + rid
+            + constant.PUBLISH_TOPIC_QUERYING_II
+            + operation_id,
+            " ",
+            callback=self.on_publish_completed,
+        )
+        self._operations[rid] = publish_action
+        self._request_response_provider.send_request(publish_action)
+
+    def _handle_response_received(self, registration_result):
+        self._query_timer.cancel()
+        self._query_timer = None
+        if self.on_registration_status_update:
+            self.on_registration_status_update(registration_result)
+
+        if registration_result.status == "assigning":
+            rid = registration_result.request_id
+            if rid in self._operations:
+                publish_action = self._operations[rid]
+                # If this request was generated by us for a query operation,
+                # then we are getting a response for a query operation and
+                # and we must clear the polling timer before beginning anything
+                if "operationId" in publish_action.publish_topic:
+                    self._polling_timer.cancel()
+                    self._polling_timer = None
+            self._trig_wait(registration_result)
+        elif registration_result.status == "assigned":
+            self._trig_complete(registration_result)
+        else:
+            self._trig_error(registration_result)
+
+    def clear_and_disconnect(self, event_data):
+        self._query_timer = None
+        self._polling_timer = None
+        unsubscribe_action = UnsubscribeAction(
+            constant.SUBSCRIBE_TOPIC_PROVISIONING, callback=self.on_unsubscribe_completed
+        )
+        self._request_response_provider.disconnect(
+            callback=self.on_disconnect_completed, pre_disconnect_actions=[unsubscribe_action]
+        )
+        # TODO : Do I need to wait for disconnect and then call the cancel callback?
+        callback = self._cancel_callback
+
+        if callback:
+            self._cancel_callback = None
+            callback()
+
+    def _call_error(self, event_data):
```

Comment:
>_call_error [](start = 8, length = 11)

If there's an error because the query timer elapsed, what happens if/when the response finally arrives?  I think it's badness

Reply:
i tried this by timing out the query timer a while back, the callback on register complete handles error (line 51 on the client)  too. In that case it just finishes the process and logs out the error.  

---
In reply to: [277409146](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277409146) [](ancestors = 277409146)

Reply:
`ERROR:azure.iot.device.provisioning.internal.polling_machine:Time is up for query timer`

---
In reply to: [277485173](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277485173) [](ancestors = 277485173,277409146)

Reply:
are you assuming that the app gets rid of the registration client and gives up?  What happens if they try again?

---
In reply to: [277486161](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277486161) [](ancestors = 277486161,277485173,277409146)

Reply:
it needs to re initialize everything again to do anything

---
In reply to: [277787698](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277787698) [](ancestors = 277787698,277486161,277485173,277409146)